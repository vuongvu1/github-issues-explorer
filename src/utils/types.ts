export type IssueType = Array<{
  id: string;
  title: string;
  number: number;
  createdAt: string;
  author: {
    login: string;
  };
  comments: {
    totalCount: number;
  };
}>;
