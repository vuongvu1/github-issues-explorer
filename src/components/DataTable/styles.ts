import styled, { css } from "styled-components";

const SC = {
  Container: styled.div(
    ({ theme }) => css`
      flex: 2;
      display: flex;
      flex-direction: column;
      padding: ${theme.spacing.md};
      background-color: ${theme.palette.white};
      box-shadow: 0px 1px 0px ${theme.palette.boxShadow};
      border-radius: 12px;
    `
  ),
  TableContainer: styled.div`
    width: 100%;
    position: relative;
  `,
  Header: styled.div(
    ({ theme }) => css`
      padding: ${theme.spacing.md};
      border: 1px solid ${theme.palette.dark};
      border-radius: ${theme.borderRadius} ${theme.borderRadius} 0 0;
      background-color: ${theme.palette.neutralLight};
      display: flex;
      align-items: center;
    `
  ),
  Row: styled.div(
    ({ theme }) => css`
      padding: ${theme.spacing.md};
      border: 1px solid ${theme.palette.dark};
      border-top: none;
      font-weight: ${theme.fontWeight.md};
      display: flex;
      align-items: center;
      justify-content: space-between;

      > * {
        &:nth-child(1) {
          align-self: flex-start;
          width: auto;
          height: 20px;
          display: flex;
          align-items: center;
          margin-right: ${theme.spacing.sm};
        }
        &:nth-child(2) {
          flex: 1;

          > a {
            text-decoration: none;
            color: ${theme.palette.black};
          }

          > * {
            &:nth-last-child(1) {
              font-size: 12px;
              font-weight: ${theme.fontWeight.sm};
              margin-top: ${theme.spacing.sm};
            }
          }
        }
        &:nth-child(3) {
          align-self: flex-start;
          width: 36px;
          display: flex;
          align-items: center;
          justify-content: space-around;
        }
      }

      &:nth-last-child(1) {
        border-radius: 0 0 ${theme.borderRadius} ${theme.borderRadius};
      }

      &:hover {
        background-color: ${theme.palette.neutralLight};
      }
    `
  ),
  LoadingOverlay: styled.div(
    ({ theme }) => css`
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      background-color: ${theme.palette.boxShadow};
      display: flex;
      justify-content: center;
      align-items: center;
    `
  ),
  Filter: styled.div<{ isActive?: boolean }>(
    ({ theme, isActive }) => css`
      display: flex;
      align-items: center;
      margin-right: ${theme.spacing.md};
      cursor: pointer;

      &:hover {
        opacity: 0.7;
      }

      ${isActive &&
      css`
        font-weight: ${theme.fontWeight.md};
      `}

      > * {
        margin-right: ${theme.spacing.sm};
      }
    `
  ),
};

export default SC;
