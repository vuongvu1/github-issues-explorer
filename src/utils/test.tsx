import { ReactNode } from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { BrowserRouter } from "react-router-dom";
import { theme } from "src/globalStyles";
import {
  IssueState,
  IssueOrderField,
  OrderDirection,
} from "src/generated/graphql";

const mockStore = configureMockStore();
const store = mockStore({
  repoSlice: {
    url: "https://github.com/reactjs/reactjs.org",
    currentPage: 1,
    searchParams: {
      name: "reactjs.org",
      owner: "reactjs",
      status: IssueState.Open,
      orderBy: IssueOrderField.CreatedAt,
      orderDir: OrderDirection.Desc,
      before: undefined,
      after: undefined,
      first: 10,
      last: undefined,
    },
  },
});

export const customRender = (Component: ReactNode) =>
  render(
    <ThemeProvider theme={theme}>
      <MockedProvider mocks={[]} addTypename={false}>
        <BrowserRouter>
          <Provider store={store}>{Component}</Provider>
        </BrowserRouter>
      </MockedProvider>
    </ThemeProvider>
  );
