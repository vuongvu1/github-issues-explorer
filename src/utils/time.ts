const ONE_YEAR_IN_SECONDS = 31536000;
const ONE_MONTH_IN_SECONDS = 2592000;
const ONE_DAY_IN_SECONDS = 86400;
const ONE_HOUR_IN_SECONDS = 3600;
const ONE_MINUTE_IN_SECONDS = 60;

export const timeSince = (date: Date) => {
  const seconds = Math.floor((new Date().getTime() - date.getTime()) / 1000);

  let interval = seconds / ONE_YEAR_IN_SECONDS;

  if (interval > 1) {
    return `${Math.floor(interval)} years`;
  }

  interval = seconds / ONE_MONTH_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} months`;
  }

  interval = seconds / ONE_DAY_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} days`;
  }

  interval = seconds / ONE_HOUR_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} hours`;
  }

  interval = seconds / ONE_MINUTE_IN_SECONDS;
  if (interval > 1) {
    return `${Math.floor(interval)} minutes`;
  }

  return `${Math.floor(seconds)} seconds`;
};
