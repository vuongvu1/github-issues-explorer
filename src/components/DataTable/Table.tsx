import React from "react";
import { useTheme } from "styled-components";
import { Link } from "react-router-dom";
import { IssueState } from "src/generated/graphql";
import { Spinner } from "src/assets/icons";
import { IssueType } from "src/utils/types";
import { timeSince } from "src/utils/time";
import {
  CircleWarning as CircleWarningIcon,
  CircleError as CircleErrorIcon,
  Tick as TickIcon,
  Comment as CommentIcon,
} from "src/assets/icons";
import SC from "./styles";

interface Props {
  issues: IssueType[];
  loading?: boolean;
  filter: IssueState;
  issueCount: number;
  totalCount: number;
  error?: string;
}

const Table: React.FC<Props> = ({
  issues,
  filter,
  loading,
  issueCount,
  totalCount,
  error,
}) => {
  const { palette } = useTheme();
  const isFilterOpen = filter === IssueState.Open;

  return (
    <SC.TableContainer>
      {loading && (
        <SC.LoadingOverlay>
          <Spinner width="48" height="48" />
        </SC.LoadingOverlay>
      )}

      <SC.Header>
        <SC.Filter isActive={isFilterOpen}>
          <CircleWarningIcon title="Open Issues" />
          {isFilterOpen ? issueCount : totalCount - issueCount} Open
        </SC.Filter>
        <SC.Filter isActive={filter === IssueState.Closed}>
          <TickIcon title="Closed Issues" />
          {!isFilterOpen ? issueCount : totalCount - issueCount} Closed
        </SC.Filter>
      </SC.Header>
      {issues.length === 0 && <SC.Row>No issues found.</SC.Row>}
      {error && <SC.Row isError>Error: {error}</SC.Row>}
      {issues.map(({ id, title, number, author, createdAt, comments }) => (
        <SC.Row key={id}>
          <div>
            {filter === IssueState.Open ? (
              <CircleWarningIcon fill={palette.success} />
            ) : (
              <CircleErrorIcon fill={palette.error} />
            )}
          </div>
          <div>
            <Link to={`/issue/${number}`}>{title}</Link>
            <div>
              #{number} opened {timeSince(new Date(createdAt))} ago by{" "}
              {author.login}
            </div>
          </div>
          <Link to={`/issue/${number}`}>
            {(comments.totalCount || null) && (
              <>
                <CommentIcon />
                &nbsp;{comments.totalCount}
              </>
            )}
          </Link>
        </SC.Row>
      ))}
    </SC.TableContainer>
  );
};

export default Table;
