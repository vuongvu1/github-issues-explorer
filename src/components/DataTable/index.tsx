import { useEffect, useState } from "react";
import { IssueState } from "src/generated/graphql";
import { IssueType } from "src/reducers/types";
import Pagination from "../Pagination";
import Table from "./Table";
import SC from "./styles";

interface Props {
  data: IssueType[];
  filter: IssueState;
  setFilter: (filter: IssueState) => void;
  loading?: boolean;
  issueCount: number;
  totalCount: number;
  error?: string;
  hasPreviousPage?: boolean;
  hasNextPage?: boolean;
  goNextPage: () => void;
  goPreviousPage: () => void;
}

const DataTable: React.FC<Props> = ({
  filter,
  setFilter,
  data,
  loading,
  issueCount,
  totalCount,
  error,
  hasNextPage,
  hasPreviousPage,
  goNextPage,
  goPreviousPage,
}) => {
  const [currentPage, setCurrentPage] = useState(1);

  const onClickPrevious = () => {
    setCurrentPage(currentPage - 1);
    goPreviousPage();
  };

  const onClickNext = () => {
    setCurrentPage(currentPage + 1);
    goNextPage();
  };

  useEffect(() => {
    setCurrentPage(1);
  }, [filter]);

  return (
    <SC.Container>
      <Table
        issues={data}
        filter={filter}
        setFilter={setFilter}
        loading={loading}
        issueCount={issueCount}
        totalCount={totalCount}
        error={error}
      />
      <Pagination
        current={currentPage}
        hasNextPage={hasNextPage}
        hasPreviousPage={hasPreviousPage}
        onClickPrevious={onClickPrevious}
        onClickNext={onClickNext}
      />
    </SC.Container>
  );
};

export default DataTable;
