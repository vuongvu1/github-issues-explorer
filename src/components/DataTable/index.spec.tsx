import { cleanup } from "@testing-library/react";
import { customRender } from "src/utils/testUtils";
import { IssueState } from "src/generated/graphql";
import DataTable from "./index";

describe("DataTable", () => {
  afterEach(cleanup);

  test("should match snapshot", () => {
    const { container } = customRender(
      <DataTable
        data={[]}
        filter={IssueState.Open}
        setFilter={() => {}}
        loading={false}
        issueCount={10}
        totalCount={20}
        error={undefined}
        hasNextPage={true}
        goNextPage={() => {}}
        hasPreviousPage={false}
        goPreviousPage={() => {}}
        currentPage={1}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
