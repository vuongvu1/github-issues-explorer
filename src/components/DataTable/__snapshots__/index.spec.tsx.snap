// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should match snapshot 1`] = `
<div>
  <div
    class="sc-iwjezw hCrrAM"
  >
    <div
      class="sc-cxpRKc iYuJDK"
    >
      <div
        class="sc-llYToB laTVUo"
      >
        <div
          class="sc-giYgFv eKVwNd"
        >
          <svg
            aria-hidden="true"
            class="prefix__octicon prefix__octicon-issue-opened"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <title>
              Open Issues
            </title>
            <path
              d="M8 1.5a6.5 6.5 0 100 13 6.5 6.5 0 000-13zM0 8a8 8 0 1116 0A8 8 0 010 8zm9 3a1 1 0 11-2 0 1 1 0 012 0zm-.25-6.25a.75.75 0 00-1.5 0v3.5a.75.75 0 001.5 0v-3.5z"
              fill-rule="evenodd"
            />
          </svg>
          10
           Open
        </div>
        <div
          class="sc-giYgFv hNVFYo"
        >
          <svg
            aria-hidden="true"
            class="prefix__octicon prefix__octicon-check"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
          >
            <title>
              Closed Issues
            </title>
            <path
              d="M13.78 4.22a.75.75 0 010 1.06l-7.25 7.25a.75.75 0 01-1.06 0L2.22 9.28a.75.75 0 011.06-1.06L6 10.94l6.72-6.72a.75.75 0 011.06 0z"
              fill-rule="evenodd"
            />
          </svg>
          10
           Closed
        </div>
      </div>
      <div
        class="sc-iJKOzS fVdXfj"
      >
        No issues found.
      </div>
    </div>
    <div
      class="sc-bdvvaa ePfWPJ"
    >
      <div>
        <button
          class="sc-gsDJrp eMQOXM"
          data-testid="prev-btn"
        >
          Previous
        </button>
        <button
          class="sc-gsDJrp gTcmdc"
        >
          1
        </button>
        <button
          class="sc-gsDJrp inDnEA"
        >
          2
        </button>
        <button
          class="sc-gsDJrp inDnEA"
          data-testid="next-btn"
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
